### ISSUE TITLE
Build fails with DISABLED_LEGACY_ENGINE

### ISSUE THREAD

**Issue by knopp** [tesseract-ocr/tesseract#3006]
### Environment

* **Tesseract Version**: 62eae84feaf9820509ec6d1395b988749753c31b
* **Commit Number**:  62eae84feaf9820509ec6d1395b988749753c31b
* **Platform**: MacOS 10.15.4

Configured with `-DSTATIC=ON -DGRAPHICS_DISABLED=ON -DDISABLED_LEGACY_ENGINE=ON -DSW_BUILD=OFF`

### Current Behavior:

Build fails: [build.log](https://github.com/tesseract-ocr/tesseract/files/4740719/build.log)

**Comment by amitdo** [comment]
I recommend to use autotools for Linux, macOS, MingGW.

```
./autogen.sh
cd build
../configure --disable-shared --disable-legacy
make -j4
./tesseract --help-psm
```

It works for me on Linux.

**Comment by knopp** [comment]
We build tesseract as dependency for product that ships both on Mac and Windows, so having consistent build would be nice. I haven't tried windows build yet, but if it the CMake build behaves same on Windows then your recommendation doesn't really help I'm afraid.

**Comment by stweil** [comment]
Using autotools provides a consistent build for all three platforms and works for me, too.

**Comment by knopp** [comment]
You're not building it with MSVC though, are you?

**Comment by stweil** [comment]
I use g++ for Windows builds.

**Comment by knopp** [comment]
Right. But that's a completely isolated ecosystem. Mingw is not ABI compatible with MSVC and can't emit PDBs either.  So for Windows projects built with MSVC (which in my experience is the case for most Winsoqa software, even though some high profile projects have been moving to clang, which is now ABI compatible)  having tesseract built with mingw is rather problematic and of limited usefulness.

**Comment by zdenop** [comment]
Build with legacy engine is working?
BTW: I doubt that DISABLED_LEGACY_ENGINE will give you any significant benefit.

**Comment by knopp** [comment]
It's working on OS X, haven't tried windows yet. The benefit is reducing final binary size about 1.1MB, which while not huge, is also not insignificant.

**Comment by zdenop** [comment]
Can you try msvc build but without `-DSTATIC=ON -DGRAPHICS_DISABLED=ON -DDISABLED_LEGACY_ENGINE=ON`?
I tried it, but it failed (but I do not try msvc often, so problem could be on my side).
MSVC build with [sw seems to compile without problem](https://ci.appveyor.com/project/zdenop/tesseract)
Build with clang (with msvc compatibility - my preferred choice) was ok (on windows 10)

**Comment by knopp** [comment]
I just tested MSVC. I get same error as I do on MacOS with DDISABLED_LEGACY_ENGINE=ON, but it builds fine with DDISABLED_LEGACY_ENGINE=OFF.

For completeness, here are the options:

-DSTATIC=ON, -DBUILD_TESTS=OFF, -DGRAPHICS_DISABLED=ON, -DSW_BUILD=OFF, -DBUILD_TRAINING_TOOLS=OFF, -DDISABLED_LEGACY_ENGINE=OFF

I'm going to make some time for this in next few days. Given that the autotools build works as expected I'll look at what the CMake build does differently.

**Comment by amitdo** [comment]
CMakeLists.txt: 

```
file(GLOB tesseract_src
    src/ccmain/*.cpp
    src/ccstruct/*.cpp
    src/ccutil/*.cpp
    src/classify/*.cpp
    src/cutil/*.cpp
    src/dict/*.cpp
    src/lstm/*.cpp
    src/opencl/*.cpp
    src/textord/*.cpp
    src/viewer/*.cpp
    src/wordrec/*.cpp
)
```

With Makefile.am the compilation of many files is prevented. For example: 

```
if !DISABLED_LEGACY_ENGINE
libtesseract_la_SOURCES += src/ccstruct/fontinfo.cpp
libtesseract_la_SOURCES += src/ccstruct/params_training_featdef.cpp
endif
```

**Comment by amitdo** [comment]
https://stackoverflow.com/questions/15550777/how-do-i-exclude-a-single-file-from-a-cmake-fileglob-pattern

**Comment by knopp** [comment]
Right. So the issue indeed seems to be that while CMakeLists.txt declares the option, it doesn't really do anything with it. The following snippet removes the legacy sources for globbed files and makes the build complete (although it doesn't do anything about legacy headers and I have no idea if the tools build).

```cmake
if (DISABLED_LEGACY_ENGINE)
    list(REMOVE_ITEM tesseract_src
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccmain/adaptions.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccmain/docqual.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccmain/equationdetect.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccmain/fixspace.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccmain/fixxht.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccmain/osdetect.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccmain/par_control.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccmain/recogtraining.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccmain/superscript.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccmain/tessbox.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccmain/tfacepp.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccstruct/fontinfo.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccstruct/params_training_featdef.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccutil/ambigs.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccutil/bitvector.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccutil/indexmapbidi.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/ccutil/universalambigs.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/adaptive.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/adaptmatch.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/blobclass.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/cluster.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/clusttool.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/cutoffs.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/featdefs.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/float2int.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/fpoint.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/intfeaturespace.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/intfx.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/intmatcher.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/intproto.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/kdtree.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/mf.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/mfdefs.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/mfoutline.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/mfx.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/normfeat.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/normmatch.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/ocrfeatures.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/outfeat.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/picofeat.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/protos.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/shapeclassifier.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/shapetable.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/tessclassifier.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/classify/trainingsample.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/dict/hyphen.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/associate.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/chop.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/chopper.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/drawfx.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/findseam.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/gradechop.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/language_model.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/lm_consistency.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/lm_pain_points.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/lm_state.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/outlines.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/params_model.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/pieces.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/plotedges.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/render.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/segsearch.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/src/wordrec/wordclass.cpp
    )
endif(DISABLED_LEGACY_ENGINE)
```

**Comment by stweil** [comment]
@knopp, is the `${CMAKE_CURRENT_SOURCE_DIR}/` prefix required? The statement which fills `tesseract_src` does not use it. As soon as the modification works, it would be nice if you could send a pull request to fix `CMakeLists.txt`.

**Comment by knopp** [comment]
The prefix is required, because `file(GLOB tesseract_src...` creates list with absolute names. So to remove the files you can either use absolute paths or match using regexp. 

The modification works, but it also needs to remove legacy header files, otherwise they get istalled. I'll at it.

**Comment by knopp** [comment]
https://github.com/tesseract-ocr/tesseract/pull/3026

**Comment by knopp** [comment]
Closed via e900252c1afac9a4088bcb3731a32e5a638a2740.
